src/java.base/share/classes/java/lang/runtime/Carrier.java

```diff
@@ -0,0 +1,921 @@
+            this.counts = CarrierCounts.count(methodType.parameterArray());
+        }
+
+        /**
+         * Return supplied methodType.
+         *
+         * @return supplied methodType
+         */
+        private MethodType methodType() {
+            return methodType;
+        }
+
+        /**
+         * Return the number of object fields needed.
+         *
+         * @return number of object fields needed
+         */
+        private int objectCount() {
+            return counts.objectCount();
+        }
+
+        /**
+         * Return the number of int fields needed.
+         *
+         * @return number of int fields needed
+         */
+        private int intCount() {
+            return counts.intCount();
+        }
+
+        /**
+         * Return the number of long fields needed.
+         *
+         * @return number of long fields needed
+         */
+        private int longCount() {
+            return counts.longCount();
+        }
+
+        /**
+         * Return parameter types.
+         *
+         * @return array of parameter types
+         */
+        private Class<?>[] ptypes() {
+            return methodType.parameterArray();
+        }
+
+        /**
+         * Return number of constructor parameters.
+         *
+         * @return number of constructor parameters
+         */
+        private int parameterCount() {
+            return methodType.parameterCount();
+        }
+
+        /**
+         * Total number of slots used in a {@link CarrierClass} instance.
+         *
+         * @return number of slots used
+         */
+        private int slotCount() {
+            return counts.slotCount();
+        }
+
+        /**
+         * Returns index of first object component.
+         *
+         * @return index of first object component
+         */
+        private int objectOffset() {
+            return 0;
+        }
+
+        /**
+         * Returns index of first int component.
+         *
+         * @return index of first int component
+         */
+        private int intOffset() {
+            return objectCount();
+        }
+
+        /**
+         * Returns index of first long component.
+         *
+         * @return index of first long component
+         */
+        private int longOffset() {
+            return objectCount() + intCount();
+        }
+    }
+
+    /**
+     * Return a constructor {@link MethodHandle} for a carrier with components
+     * aligning with the parameter types of the supplied
+     * {@link MethodType methodType}.
+     *
+     * @param methodType  {@link MethodType} providing types for the carrier's
+     *                    components
+     *
+     * @return carrier constructor {@link MethodHandle}
+     *
+     * @throws NullPointerException is any argument is null
+     * @throws IllegalArgumentException if number of component slots exceeds maximum
+     */
+    public static MethodHandle constructor(MethodType methodType) {
```

Comment:
What happens to the methodType return type? (both here and in the components method). Should javadoc say anything?

Reply:
The return type is ignored. Going back to John's notion of just passing in the parameterTypes would work okay. I think when Brian wrote the original code he may have been thinking about keying on the MethodType (interned). I'll add a note about the return type.